### DISCUSSION TITLE
FHIR Operations in AccessPolicy

### DISCUSSION THREAD

**Discussion by codyebberson** [medplum/medplum#4374] [Ideas]
## Background

Medplum has a very powerful `AccessPolicy` feature, which can be used to control access to FHIR Resources.

Some quick examples of `AccessPolicy` features:

1. Which resource types the user has access to
2. Whether resource types are read only or read/write
3. Which elements (properties) a user can edit
4. Which elements (properties) a user can see
5. Write constraints, which act as post-edit pre-commit checks on a resource

One area that is missing from the current `AccessPolicy` implementation is [FHIR Operations](https://hl7.org/fhir/R4/operations.html).

> The [RESTful API](https://hl7.org/fhir/R4/http.html) defines a set of common interactions (read, update, search, etc.) performed on a repository of typed resources. These interactions follow the RESTful paradigm of managing state by Create/Read/Update/Delete actions on a set of identified resources. While this approach solves many use cases, there is some functionality that can be met more efficiently using an RPC-like paradigm, where named operations are performed with inputs and outputs (Execute).

This all becomes even more powerful considering that Medplum specific entities are represented as FHIR Resources (`Project`, `User`, `ProjectMembership`, etc), and many non-FHIR Medplum APIs are represented as FHIR Operations (`Agent/$push`, `Bot/$execute`, etc)

### Backwards compatibility vs breaking changes

The `AccessPolicy` structure has been stable for a while, by design.  Users generally do not like to modify security settings, as there is always the inherent risk of inadvertently breaking things when you touch it.

We should be sensitive and respectful of user needs when making changes to `AccessPolicy`.

When considering backwards compatibility vs breaking changes, we should clearly choose one:

1. An incremental change which is backwards compatible with the existing implementation
2. A clear breaking change, which is 10x better than the current implementation, so the benefits outweigh the costs

### Interaction vs Operation

We have several internal conversations going on about aligning "interaction" and "operation" into a unified framework within the Medplum server implementation.

Currently, these are implemented in somewhat ad hoc fashion, typically as routes on a router.  But that means that some aspects such as permissions and logging are implemented ad hoc as well.

There may be an elegant option here where "interaction" and "operation" are both represented together in `AccessPolicy` in a unified way.

### AccessPolicy inheritance

Currently, each AccessPolicy is completely independent, no inheritance of includes or `partOf` or `link` or anything like that.

There is a very clear and desirable feature, to be able to reference other AccessPolicies, especially for something like "IP access rules", which enterprise customers often manage globally rather than per "group" or user type.

How would operations fit in with that?

### "Resource.resourceType" vs "Element.path"

Tangentially related to this change, in the past we have discussed making `AccessPolicy` centered on `Element` and `ElementDefinition` rather than `Resource`.  That has some elegant properties to it, such as aligning with `StructureDefinition` and the FHIR spec in general.  But it would probably also move us into "breaking changes" territory, for unclear gains.

I think the "Resource" vs "Element" change is probably out of scope for this discussion.  I only bring it up so that we can evaluate how the solution to FHIR Operations may be forward compatible with that change in the future.

### Super dumb and simple: `AccessPolicy.resource.allowedOperations`

We have an existing backbone element on `AccessPolicy.resource` which represents access controls for a given resource type:

```ts
/**
 * Access details for a resource type.
 */
export interface AccessPolicyResource {

  /**
   * The resource type.
   */
  resourceType: string;

  /**
   * The rules that the server should use to determine which resources to allow.
   */
  criteria?: string;

  /**
   * Optional flag to indicate that the resource type is read-only.
   */
  readonly?: boolean;

  /**
   * Optional list of hidden fields.  Hidden fields are not readable or writeable.
   */
  hiddenFields?: string[];

  /**
   * Optional list of read-only fields.  Read-only fields are readable but not writeable.
   */
  readonlyFields?: string[];

  /**
   * Invariants that must be satisfied for the resource to be written.  Can
   * include %before and %after placeholders to refer to the resource
   * before and after the updates are applied.
   */
  writeConstraint?: Expression[];
}
```

We could add a new element:

```ts
  /**
   * Optional list of operations the user has access to.
   */
  allowedOperations?: string[];
```

Pros:
 * Easy to understand
 * Easy to implement

Cons:
 * Unwieldy to maintain?  The list of allowed operations could be long (20+ entries), which may be a lot of copy pasting
 * Does not provide any granularity within an operation

### Medium: Operation Rules

Similar to the previous option, we could add a new element `operationRules` and a new type `AccessPolicyOperationRule`.

The model would use a linear list of rules, existing at the first matching rule.

* allow or block - 
* resource constraints - FHIRPath expression on the root resource
* parameters constraints - FHIRPath expression on the input `Parameters`

Pros:
* Expressive
* Moderately easy to implement

Cons:
* More verbose
* More difficult to author?  Not sure if the extra power is worth the complexity?

### Others?

I know Medplum team members and community members have ideas here.  Please share!

**Comment by rahul1** [comment]
> resource constraints - FHIRPath expression on the root resource
> parameters constraints - FHIRPath expression on the input Parameters

@codyebberson Did you have any use cases off the top of your head for these types of constraints? Mostly curious about the decision to go with FHIRPath expression for resource constraints vs. search expressions. One argument I could see for that is to put less pressure on our in-memory search implementation. 

One possible resolution to the unwieldiness of `AccessPolicy.resource.allowedOperations` is to make two fields, and allow and a block list: 
- `AccessPolicy.resource.allowedOperations`
- `AccessPolicy.resource.blockedOperations`

We'd have to define a precedence, but we can maybe leverage some best practices from [AWS IAM](https://docs.aws.amazon.com/IAM/latest/UserGuide/reference_policies_evaluation-logic.html) (they have deny > allow).
